canonical-web-and-design/maas-ui
# refactor(ui): convert power utils to TS and move to general directory

## Done

- Converted power utils to TS and moved to `app/store/general/utils`:
  - `formatPowerParameters` has changed args so that chassis and scope are separate (necessary for machine config)
  - `usePowerParametersSchema` is now `generatePowerParametersSchema` (no longer a hook). It generates a Yup object shape based on the given power type and its fields
  - `useAllPowerParameters` is now `useInitialPowerParameters` - I'll need to be able to initialise the form with predefined values for the machine configuration form so it's been updated to allow that

## QA

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Karura](/HACKING.md#karura)
- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps

- Check that the forms for Add Machine, Add Chassis and Add KVM all still validate properly
- Try to add a machine (doesn't have to be valid values) and check the websocket request - make sure it only includes the values relevant to the selected power type

## Fixes

Fixes canonical-web-and-design/MAAS-squad#2333
